### ISSUE TITLE
FR: Reduce rainbow effect on Kobo Colour

### ISSUE THREAD

**Issue by Blendman974** [koreader/koreader#11877]
**Does your feature request involve difficulty completing a task? Please describe.**
Kaleido 3 color displays (such as Kobo Libra Colour and Clara Colour) tend to produce a "rainbow effect" when displaying certain image patterns. 
Kobo's color readers have a "reduce rainbow effect" toggle, which appears to add a slight blur to images before displaying them. This can sometimes help with the rainbow effect but also makes the images somewhat blurry.

**Describe the solution you'd like**
Implement the same feature found in the default Kobo reader software, a "reduce rainbow effect"

**Describe alternatives you've considered**
Discussions online suggest that the rainbow effect is caused by specific patterns, and that more advanced algorithms could potentially eliminate the effect entirely. 
However, these algorithms are more complex and require intensive computation, making them more suitable for other projects like kcc that process files on a computer.

**Additional context**
Here are [images of the rainbow effect](https://www.reddit.com/r/kobo/comments/1chs0kh/i_think_i_found_the_rainbow_effect_libra_colour/)

**Comment by Frenzie** [comment]
In developer options (in the file browser, Tools → More tools → Developer options) you can play with hardware/software/no dithering btw.

**Comment by Blendman974** [comment]
Tinkering with the dithering settings in the developer options hasn't made any difference. The rainbow issue seems to be a hardware limitation related to how the color screen functions. 
Certain image patterns, regardless of what you do, will produce this effect. 
I don't know exactly what the "reduce rainbow effect" toggle does on the stock reader, but people online suggest that it applies a slight blur or smoothing to the image, which helps eliminate the problematic patterns.

**Comment by Frenzie** [comment]
Maybe something like this: https://github.com/RayXie29/Chromatic_aberration_correction

**Comment by Blendman974** [comment]
My understanding is that the effect is created by specific pixel patterns on the image. This is explained by a comment on Reddit: [link to the comment](https://www.reddit.com/r/kobo/s/nXvoLc5Qp4).

Additionally, chromatic aberration only occurs on color content so I don't think it will have an impact.
The "reduce rainbow" toggle likely smooths the image to remove the problematic pattern.

This can be verified by reducing the resolution of the images or blurring them, which causes the effect to disappear.
Ideally, we should determine the exact amount of blur added by the stock reader and apply the same. 

In the future, a more advanced solution could be developed to detect and smooth only the problematic patterns without affecting the rest of the image, similar to techniques used for removing chromatic aberration. But those are way above my understanding.

**Comment by Frenzie** [comment]
It might be worth noting that if you set the zoom setting to 300 you'll get slightly blurrier pictures (and DjVu/PDF you're probably reading with some resizing regardless). It's not done by default for some reason or other (presumably mainly performance and/or not making things blurrier) but I'd say there's a decent chance you want to do so on modern devices regardless to avoid tiny pictures.

But long story short maybe added blurriness inherent in the resizing process will help against this effect?
<img src=https://github.com/koreader/koreader/assets/202757/cfa968cf-5c85-47aa-b37c-3258fc87327c width=40%>

**Comment by offset-torque** [comment]
Manga books generally use ordered dither patterns to create gray as in the example. This is from Battle Angel Alita, one of the books linked as problematic in the Reddit thread:
![alita_ordered_dither](https://github.com/koreader/koreader/assets/84420582/7de1276f-d813-4a5c-9332-9976d54e527d)

When showing images like this, dither pattern sometimes overlaps with the frequency of the Kaleido Color Filter Array in a staggering pattern and creates the rainbow.
![kaleido_cfa](https://github.com/koreader/koreader/assets/84420582/bc3dc043-8b93-48a3-854f-8eb1d01b6d4f)

Blurring or resizing works by breaking this overlapping interference. I think blurring might be better because resizing can shift this pattern and can create this effect on some other areas.

**Comment by NiLuJe** [comment]
Yup, the actual physical CFA layer produces is extremely prone to Moiré. And as it's literally a physical layer, we can't do much about it (e.g., switching to grayscale won't help).

Nickel pulled in OpenCV to add an optional pass that does... something (it does nothing on my end, or at least did at the time ;p), which I guess is probably some sort of blur.

I have zero interest in doing that, but patches welcome.

**Comment by Blendman974** [comment]
The blur on nickel is only done on kepub files and seems very noticeable on my end.

You can actually break the grid patterns on the image by making changes on the bright spots of the image Fourier transform.

Currently trying to create an implementation of this algorithm.
There are multiple papers about it online.
This one is in french but is almost what I'm trying to implement: https://clouard.users.greyc.fr/Pantheon/experiments/scanner-restoration/index-fr.html

**Comment by AryaTwirl** [comment]
@Blendman974 Hi, I've read the link you've posted already, but I can't quite grasp how I can do that to my already existing CBZ file. Is that using something called Pandora?

**Comment by radugrecu97** [comment]
@Blendman974 Could you please share your algorithm?

**Comment by radugrecu97** [comment]
@Blendman974

TL;DR I achieved greatly reducing the rainbow pattern while maintaining the crisp sharpness. Tested on Kobo Libra Color.

-----------------------

 I succeeded using Pandore to process images but it was too slow on my rapsberry pi.

So then I managed to replicate the Pandore functionality in Python with a big increase in speed and then parallelized the processing.

The main issue is that the initial output loses contrast, whites become grey, blacks are less black. So I normalize them again by referencing the input image white and black pixel locations.

It worked really well and the final output, while still noticeable, has a definite decrease in rainbows (at least on the Jujutsu Kaisen prints, need to test others). But not complete removal!

I'll try to come up with a better strategy to aim for removal of the problem areas. I think it should ideally blend the screentone (the dotted pattern used in manga shading) so it's one solid color.

Repo can be found here:
https://github.com/radugrecu97/manga_downloader_and_sticher

You only need to use the remove_moiree.py

**Comment by rukit24** [comment]
@radugrecu97 
I have test with your work on my Jujutsu Kaisen Raw, and did you feel it was made the image too white and seem to blur. And somehow japanese typo just scratch and difficult to read. 
By the way thanks for your fantasic work and hoping it will be completed in the future

**Comment by radugrecu97** [comment]
> [@radugrecu97](https://github.com/radugrecu97) I have test with your work on my Jujutsu Kaisen Raw, and did you feel it was made the image too white and seem to blur. And somehow japanese typo just scratch and difficult to read. By the way thanks for your fantasic work and hoping it will be completed in the future

I also used HQ volume scans without any blur in the output. Do you have "remove rainbow effect" disabled if you use epub? That blurs the image.

Can you upload a before and after?

**Comment by rukit24** [comment]
> I also used HQ volume scans without any blur in the output. Do you have "remove rainbow effect" disabled if you use epub? That blurs the image.
> 
> Can you upload a before and after?

actually my workflow is that i am using Mangajanai to upscale raw to 4k then using some code working with Mapick to resize it to kobo libra size. then i am apply your code to handle the rainbow issue. So your workflow is apply the pandore then packing in KCC?

**Comment by radugrecu97** [comment]
> > I also used HQ volume scans without any blur in the output. Do you have "remove rainbow effect" disabled if you use epub? That blurs the image.
> > Can you upload a before and after?
> 
> actually my workflow is that i am using Mangajanai to upscale raw to 4k then using some code working with Mapick to resize it to kobo libra size. then i am apply your code to handle the rainbow issue. So your workflow is apply the pandore then packing in KCC?

My only guess is that the upscaling algorithm interferes with moiree removal. I'll double check later if my repo is up-to-date.

I'm downloading from weebcentral because it has the highest quality I've seen but it's only chapters and not volumes.

I download all the relevant chapters. Apply moiree fix.

Then parse Wikipedia to correctly pack the chapters into volumes. Download covers from another website automatically.

KCC cli to make separate volumes.

**Comment by Its-my-right** [comment]
@Blendman974 thanks you a lot for the Fourier Transform Idea !
I managed to create a .so library that removes on the fly the rainbow effect on Koreader (this way, you do not have to alter the pictures before opening it via koreader).

It took a long time because it was my first C code (I know Python but did not know well C before that), and the major part of this was done with the help of AI (chatGpt a bit but mostly Claude sonnet 3.7).

I have 3 versions of the code because I tried with 3 differents FFT libraries: FFTW, FFTS and kiss-fft. The best result (process time) is achieved with FFTW but I am not sure if the licensing of FFTW would permit to use it officialy in koreader or not.

The added processing time is under 400ms per page with FFTW, so it is totally acceptable for reading. And maybe it could be improved more (But I already managed to decrease it from 1,5s to < 400ms with the help of Claude).

I load the .so library via a Koreader lua patch (userpatch module), and it applies only on BW pictures.

I can provide the .so binary + lua patch if some of you want to try. Of course I can also provide source code but I totally suck at github etc so not sure how to do it.

But I will need to provide other stuff too, because of how my full process works:
- It starts with a lua patch that loads 3 .so module I wrote/compiled: module1 is for checking if current framebuffer is color or BW, module2 is for removing moire via Fourier Transform on BW images, module3 is for improving colors on color images
- The lua patch then calls these modules to apply processes on bw and colored images

The lua patch comes with some parameters to adjust if necessary.

The great aspect of using Fourier Transform instead of the usual blurring method is that you nearly don't lose useful details (of course the method breaks patterns in the pictures so you factually lose details but it is nearly invisible and unlike the blurring method, texts and edges remains crisp)

An other important information: the .so modules (and C code associated) were wrotten with the arm cortex a7 in mind (because my Pocketbook Inkpad Color 3 uses this cpu). Maybe the code will need a bit of rewrite to me more generic or to adapt to other cpus/readers.

Anyway all of this is fully functionnal on Koreader on my Pocketbook Inkpad Color 3 so if you have this exact setup it should work perfectly too, and on another setup it could need some light rewrite.

Here is a video comparison of the same page displayed without an with my library: https://youtube.com/shorts/8rw_-P6NSEk?feature=share

**Comment by rukit24** [comment]
> [@Blendman974](https://github.com/Blendman974) thanks you a lot for the Fourier Transform Idea ! I managed to create a .so library that removes on the fly the rainbow effect on Koreader (this way, you do not have to alter the pictures before opening it via koreader).
> 
> It took a long time because it was my first C code (I know Python but did not know well C before that), and the major part of this was done with the help of AI (chatGpt a bit but mostly Claude sonnet 3.7).
> 
> I have 3 versions of the code because I tried with 3 differents FFT libraries: FFTW, FFTS and kiss-fft. The best result (process time) is achieved with FFTW but I am not sure if the licensing of FFTW would permit to use it officialy in koreader or not.
> 
> The added processing time is under 400ms per page with FFTW, so it is totally acceptable for reading. And maybe it could be improved more (But I already managed to decrease it from 1,5s to < 400ms with the help of Claude).
> 
> I load the .so library via a Koreader lua patch (userpatch module), and it applies only on BW pictures.
> 
> I can provide the .so binary + lua patch if some of you want to try. Of course I can also provide source code but I totally suck at github etc so not sure how to do it.
> 
> But I will need to provide other stuff too, because of how my full process works:
> 
> * It starts with a lua patch that loads 3 .so module I wrote/compiled: module1 is for checking if current framebuffer is color or BW, module2 is for removing moire via Fourier Transform on BW images, module3 is for improving colors on color images
> * The lua patch then calls these modules to apply processes on bw and colored images
> 
> The lua patch comes with some parameters to adjust if necessary.
> 
> The great aspect of using Fourier Transform instead of the usual blurring method is that you nearly don't lose useful details (of course the method breaks patterns in the pictures so you factually lose details but it is nearly invisible and unlike the blurring method, texts and edges remains crisp)
> 
> An other important information: the .so modules (and C code associated) were wrotten with the arm cortex a7 in mind (because my Pocketbook Inkpad Color 3 uses this cpu). Maybe the code will need a bit of rewrite to me more generic or to adapt to other cpus/readers.
> 
> Anyway all of this is fully functionnal on Koreader on my Pocketbook Inkpad Color 3 so if you have this exact setup it should work perfectly too, and on another setup it could need some light rewrite.
> 
> Here is a video comparison of the same page displayed without an with my library: https://youtube.com/shorts/8rw_-P6NSEk?feature=share

Great jobs, can you share your works. I will make a test.

**Comment by GlazedBelmont** [comment]
Interested in testing too

**Comment by offset-torque** [comment]
Hey folks, I want to remind you that once you reach a working or satisfactory solution I can add it to the user guide. So instead of being buried in this issue, other people can benefit from your efforts too. We already have a "Optimizing manga for reader" section in the guide for example. This would be a nice complement, considering that color e-ink is getting cheaper and more widespread.

Of course if you have a github page/gist/repo or own webpage documenting this, I can just add a link to it too.

**Comment by Its-my-right** [comment]
Hi,

I quiclky created my first repo with all necessary files and explanations:

https://github.com/Its-my-right/pocketbook_cfa_interference_breaker

If you have a Pocketbook Inkpad color 3, you can directly use the binaries provided here :)

For other e-reader, I recommend modifiyng the sources and re-compiling.

Let me know how it works for you :)

**Comment by rukit24** [comment]
> Hi,
> 
> I quiclky created my first repo with all necessary files and explanations:
> 
> https://github.com/Its-my-right/pocketbook_cfa_interference_breaker
> 
> If you have a Pocketbook Inkpad color 3, you can directly use the binaries provided here :)
> 
> For other e-reader, I recommend modifiyng the sources and re-compiling.
> 
> Let me know how it works for you :)

I copy your modules_for_pocketbook_inkpad_color_3 inside folder into my Kobo Libra Colour Koreader folder but it has error : "Error to load patch". i saw your folder has folder custom_libs meanwhile koreader folder is libs, so i move file to libs folder but doesn't work. Does this because my Ereader using different CPU than your? Kobo Libra Colour uses ARMv7

**Comment by Its-my-right** [comment]
> > Hi,
> > I quiclky created my first repo with all necessary files and explanations:
> > https://github.com/Its-my-right/pocketbook_cfa_interference_breaker
> > If you have a Pocketbook Inkpad color 3, you can directly use the binaries provided here :)
> > For other e-reader, I recommend modifiyng the sources and re-compiling.
> > Let me know how it works for you :)
> 
> I copy your modules_for_pocketbook_inkpad_color_3 inside folder into my Kobo Libra Colour Koreader folder but it has error : "Error to load patch". i saw your folder has folder custom_libs meanwhile koreader folder is libs, so i move file to libs folder but doesn't work. Does this because my Ereader using different CPU than your? Kobo Libra Colour uses ARMv7

The custom_libs directory is correct.
The problem here mostly is in the lua patch: Kobo framebuffer is different from pocketbook framebuffer, so the patch needs to be tweaked to adapt to Kobo framebuffer.

As stated on the readme, this module works directly if you have a Pocketbook InkPad Color 3, but needs to be adapted if you have an other reader (especially an other brand)

**Comment by Moxiesan** [comment]
Has anyone attempted updating this to work for a Colorsoft?

**Comment by axu2** [comment]
The algorithm is included in kindle comic converter 9 as the rainbow eraser by @Its-my-right

**Comment by axu2** [comment]
In case it's useful, I made a quick overview video about the math theory behind rainbow erasing on Kaleido 3 color eink (if I made any technical mistakes let me know, I only recently learned about the math): 

KCC users really like the feature.

Fix rainbows on color eink with the Discrete Fourier Transform!
https://youtu.be/Dw2HTJCGMhw?si=A-vSylmROPHznsZl

**Comment by Frenzie** [comment]
I was distracted by the pronunciation of FOO-ree-ay. :-P

Cool video. 👍

**Comment by Euphoriyy** [comment]
I have made a plugin called [derainbowify.koplugin](https://github.com/Euphoriyy/derainbowify.koplugin) based on the patch shared by @Its-my-right that allows you to use to remove the rainbow effect directly on your device using the aforementioned method of fast fourier transforms.

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/4feff5e3-5618-410e-b3de-7a51431b97e0" />

---

Currently it has been tested for the Kobo Libra Color (the only e-reader I own), although IIRC the Kindle Colorsoft should be able to use it as well. However, after I automate the build process it should be trivial to support the Pocketbook Inkpad Color 3 and every other device.

It's a bit slow when loading a page for the first time, but I'm currently using PFFFT with a single thread instead of FFTW with multiple threads on the Inkpad Color 3, so that might explain it. Anyways I'll try to improve the performance as much as possible.

If anyone has any feedback, especially with testing on specific devices, please let me know!

Thanks so much to @Blendman974 for the idea and to @Its-my-right for doing the heavy lifting!

---

EDIT: I've added the derainbow effect to pre-rendered pages (hinting), and it works so much better now in practice with manga/comics. No noticeable delay even without multi-threading.

**Comment by Euphoriyy** [comment]
With version 0.0.2 of the plugin, I've now added cross-platform support for Kindle and Pocketbook as well. If anyone owns those devices and is interested, please test the plugin out and leave me feedback as to how well it works! 😄

**Comment by leodbrs** [comment]
I’ve been using this plugin on my Kobo Libra Colour for three months, and I’ve never had any issues. Thank you for it

**Comment by offset-torque** [comment]
Your plugin looks really useful @Euphoriyy
I am adding your plugin to the `READING COMICS AND MANGA` section of our guide.
There is enough information in your repo README it seems. But if there is something you need to add/clarify there, I suggest updating it because you will probably get more users after the guide update.

**Comment by Euphoriyy** [comment]
> I’ve been using this plugin on my Kobo Libra Colour for three months, and I’ve never had any issues. Thank you for it

That means so much to me! I'm glad you were able to find benefit from my work and that it worked nicely. I appreciate your feedback. 😁 

> Your plugin looks really useful [@Euphoriyy](https://github.com/Euphoriyy) I am adding your plugin to the `READING COMICS AND MANGA` section of our guide. There is enough information in your repo README it seems. But if there is something you need to add/clarify there, I suggest updating it because you will probably get more users after the guide update.

Thanks! I appreciate your hard work on the user guide. A bit late of me, but I've just clarified details about the performance, fixed some wording, and linked directly to the releases page.
